flang/lib/Evaluate/tools.cpp

```diff
@@ -1210,6 +1210,26 @@ bool HasConstant(const Expr<SomeType> &expr) {
   return HasConstantHelper{}(expr);
 }
 
+// HasStructureComponent()
+struct HasStructureComponentHelper
+    : public AnyTraverse<HasStructureComponentHelper, bool, false> {
+  using Base = AnyTraverse<HasStructureComponentHelper, bool, false>;
+  HasStructureComponentHelper() : Base(*this) {}
+  using Base::operator();
+
+  bool operator()(const DataRef &dataRef) const {
+    return std::holds_alternative<Component>(dataRef.u);
+  }
+  bool operator()(const NamedEntity &NamedEntity) const {
+    return NamedEntity.UnwrapComponent() != nullptr;
+  }
+  bool operator()(const Component &) const { return true; }
```

Comment:
Do we need the other two visitors or will it still catch them here after one more traversal?